Travis: define matrix builds for mbed 2 jobs

0xc0170 opened the pull request:
Use matrix for our targets, to get smaller jobs that can run in parallel.

0xc0170 commented:
I think I have it now defined as initially intended. We still use matrix, use env to set vendor to run tests for (to avoid duplication, same matrix job just different env, to set up multiple jobs). We were over the limit of 50 minutes. As travis build script runs, its build libs, then links examples.

It is the same, just that one job per targets/TARGET_xxxx folder (named vendor in the travis script). The latest job should pass!

0xc0170 commented:
/morph build

0xc0170 commented:
Travis fails for SAM devices,  10 minutes no log, investigating locally

geky commented on .travis.yml:
```diff
@@ -34,6 +34,8 @@ after_success:
 after_failure:
   - bash -c "$STATUS" failure "Local $NAME testing has failed"
 
+language: generic
```
Hmm, should this be python?

geky commented on .travis.yml:
```diff
@@ -177,6 +230,55 @@ matrix:
+
+    - python: '2.7'
+      env:
+        - NAME=mbed2-NUVOTON
```
I bet there's a way to not have to repeat yourself here... but I tried it on my fork and ran into issues.

Will look into this later...

0xc0170 commented:
Still atmel job timeouts. I run it locally takes 460s, run through profiler, multiple times, could not reproduce the timeout (however travis runs multiple jobs (3 at the time)).

geky commented:
Looks like it was an issue with gcc-arm-none-eabi version.

gcc version 7.1 added `-Wexpansion-to-defined`, which warns if a macro expands to contain `defined(blablabla)`:
https://www.gnu.org/software/gcc/gcc-7/changes.html

Macros containing defined were used **heavily** in the Atmel HAL. Things exploded to the point where build.py died. This is probably the reason we started seeing the mbed 2 timeouts in the first place.

700f06e Removes the warning. There's not a clear workaround for nested macro defineds and they are pretty entrenched in the Atmel HAL.

0xc0170 commented:
/morph build

0xc0170 commented:
I updated the travis file, to specify version (fetch it from official arm website), to stay with GCC ARM 6 as we currently support. For Atmel fix, we will create a new tracking issue wit hthe provided fix here , that needs further investigation. To stay compatible, I reverted that change here (rebased).

thanks @geky for the clean-up, looks good !

0xc0170 commented:
/morph build

0xc0170 commented:
I found one mistype in yaml file that I fixed, restarting CI, once done this should be ready to get in

0xc0170 commented:
This is now ready for integration 💯
